### PR TITLE
Use vue-tsc instead of tsc

### PR DESCRIPTION
**Pull request by JammingBen**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Technical debt
- [ ] Tests

## Followup Tasks (in order)
- [ ] Register ODS components with GlobalComponents and fix issues
- [ ] Port all mixins to composables (or remove obsolete ones)
- [ ] Enable `compilerOptions.noImplicitThis`, c.f. https://vuejs.org/guide/typescript/overview.html#configuring-tsconfig-json
       and fix remaining issues

### CODE REVIEW

**Comment by dschmidt** [comment]
... some of the changes are actual fixes, some are just shitty workaround :-\

The output of vue-tsc is a bit hard to read, so we kept the workarounds to make the output more readable ...
We should pick the fixes and file issues for the workarounds at the vue-tsc repo.

File: packages/design-system/src/components/OcSearchBar/OcSearchBar.vue [hunk 1]
```diff
@@ -21,7 +21,7 @@
         :value="searchQuery"
         :disabled="loading"
         :placeholder="placeholder"
-        @input="onType($event.target.value)"
+        @input="onType(($event.target as any).value)"
```
Comment (dschmidt) [inline packages/design-system/src/components/OcSearchBar/OcSearchBar.vue:24]:
```suggestion
        @input="onType(($event.target as HTMLInputElement).value)"
```

File: packages/design-system/src/components/OcTextInput/OcTextInput.vue [hunk 2]
```diff
@@ -15,8 +15,8 @@
         :type="type"
         :value="displayValue"
         :disabled="disabled"
-        @change="onChange($event.target.value)"
-        @input="onInput($event.target.value)"
+        @change="onChange(($event.target as any).value)"
+        @input="onInput(($event.target as any).value)"
```
Comment (dschmidt) [inline packages/design-system/src/components/OcTextInput/OcTextInput.vue:19]: HTMLInputElement

**Comment by dschmidt** [comment]
See also https://github.com/vuejs/language-tools/pull/2471

File: packages/design-system/src/components/OcResource/OcResource.vue [hunk 3]
```diff
@@ -105,7 +105,7 @@ export default defineComponent({
      * The resource to be displayed
      */
     resource: {
-      type: Object,
+      type: Object as PropType<any>,
```
Comment (dschmidt) [inline packages/design-system/src/components/OcResource/OcResource.vue:108]: Why didn't we use `Resource`?

File: packages/web-app-admin-settings/src/components/Users/SideBar/DetailsPanel.vue [hunk 4]
```diff
@@ -130,6 +130,9 @@ export default defineComponent({
         .sort()
         .join(', ')
     },
+    userGroupLength() {
+      return this.user.memberOf.length
+    },
```
Comment (dschmidt) [inline packages/web-app-admin-settings/src/components/Users/SideBar/DetailsPanel.vue:135]: hack

File: packages/web-app-admin-settings/src/components/Users/SideBar/EditPanel.vue [hunk 5]
```diff
@@ -214,6 +214,9 @@ export default defineComponent({
       return this.$gettext(
         this.roles.find((role) => role.id === assignedRole?.appRoleId)?.displayName || ''
       )
+    },
+    userId() {
+      return this.user.id
```
Comment (dschmidt) [inline packages/web-app-admin-settings/src/components/Users/SideBar/EditPanel.vue:219]: hack

File: packages/web-app-files/src/components/AppBar/AppBar.vue [hunk 6]
```diff
@@ -76,14 +76,17 @@ import SharesNavigation from './SharesNavigation.vue'
...
 export default defineComponent({
   components: {
     BatchActions,
     ContextActions,
     SharesNavigation,
     SidebarToggle,
-    ViewOptions
+    ViewOptions,
+    // Remove when mixins are replaced
+    ...ODSComponents
```
Comment (dschmidt) [inline packages/web-app-files/src/components/AppBar/AppBar.vue:89]: hack: we should be able to get rid of this with action composables: https://github.com/owncloud/web/pull/8542

File: packages/web-app-files/src/components/AppBar/CreateAndUpload.vue [hunk 7]
```diff
@@ -179,10 +173,13 @@ import { stringify } from 'qs'
...
 import { UppyService } from 'web-runtime/src/services/uppyService'
 import { getIndicators } from 'web-app-files/src/helpers/statusIndicators'
+import * as ODSComponents from '@ownclouders/design-system/src/components'
 
 export default defineComponent({
   components: {
-    ResourceUpload
+    ResourceUpload,
+    // Remove when mixins are replaced
+    ...ODSComponents
```
Comment (dschmidt) [inline packages/web-app-files/src/components/AppBar/CreateAndUpload.vue:182]: hack

File: packages/web-app-files/src/components/Search/List.vue [hunk 8]
```diff
@@ -167,6 +167,9 @@ export default defineComponent({
         itemCount: this.itemCount,
         totalResults: this.searchResult.totalResults
       })
+    },
+    routeQueryTerm() {
+      return this.$route.query.term
```
Comment (dschmidt) [inline packages/web-app-files/src/components/Search/List.vue:172]: Was there any reason for this?

File: packages/web-app-files/src/components/Search/Preview.vue [hunk 9]
```diff
@@ -36,10 +36,15 @@ import { buildShareSpaceResource, Resource } from 'web-client/src/helpers'
...
 import { eventBus } from 'web-pkg/src/services/eventBus'
 import { createFileRouteOptions } from 'web-pkg/src/helpers/router'
+import * as ODSComponents from '@ownclouders/design-system/src/components'
 
 const visibilityObserver = new VisibilityObserver()
 
 export default defineComponent({
+  components: {
+    // Remove when mixins are replaced
+    ...ODSComponents
```
Comment (dschmidt) [inline packages/web-app-files/src/components/Search/Preview.vue:46]: hack

File: packages/web-app-files/src/components/Shares/SharedWithMeSection.vue [hunk 10]
```diff
@@ -118,7 +119,9 @@ export default defineComponent({
     ResourceTable,
     ContextActions,
     ListInfo,
-    NoContentMessage
+    NoContentMessage,
+    // Remove when mixins are replaced
+    ...ODSComponents
```
Comment (dschmidt) [inline packages/web-app-files/src/components/Shares/SharedWithMeSection.vue:124]: hack

File: packages/web-app-files/src/components/SideBar/Actions/FileActions.vue [hunk 11]
```diff
@@ -16,10 +16,15 @@ import ActionMenuItem from 'web-pkg/src/components/ContextActions/ActionMenuItem
...
 import { Resource, SpaceResource } from 'web-client'
+import * as ODSComponents from '@ownclouders/design-system/src/components'
 
 export default defineComponent({
   name: 'FileActions',
-  components: { ActionMenuItem },
+  components: {
+    ActionMenuItem,
+    // Remove when mixins are replaced
+    ...ODSComponents
```
Comment (dschmidt) [inline packages/web-app-files/src/components/SideBar/Actions/FileActions.vue:26]: hack

File: packages/web-app-search/src/views/List.vue [hunk 12]
```diff
@@ -10,7 +10,7 @@ import { debounce } from 'lodash-es'
 
 export default defineComponent({
   data() {
-    const { provider: providerId } = this.$route.query
+    const { provider: providerId } = (this.$route as any).query
```
Comment (dschmidt) [inline packages/web-app-search/src/views/List.vue:13]: why don't we cast `.query` instead?

File: packages/web-runtime/src/components/Topbar/Notifications.vue [hunk 13]
```diff
@@ -102,6 +103,21 @@ const POLLING_INTERVAL = 30000
...
+    const { owncloudSdk: client } = useClientService()
+    return { client }
+  },
+  computed: {
+    ...mapGetters(['activeNotifications', 'configuration']),
+
+    notificationsLabel() {
+      return this.$gettext('Notifications')
+    }
+>>>>>>> f8e9dd126 ('Fix' more ts type errors in vue files)
```
Comment (dschmidt) [inline packages/web-runtime/src/components/Topbar/Notifications.vue:120]: messed up rebase

File: packages/web-runtime/src/components/Topbar/ThemeSwitcher.vue [hunk 14]
```diff
@@ -24,9 +24,10 @@ export default defineComponent({
     const currentTheme = computed(() => store.getters.configuration.themes[unref(currentThemeName)])
     const applyTheme = (theme) => {
       for (const param in theme.designTokens.colorPalette) {
-        document
-          .querySelector(':root')
-          .style.setProperty(`--oc-color-${param}`, theme.designTokens.colorPalette[param])
+        ;(document.querySelector(':root') as any).style.setProperty(
```
Comment (dschmidt) [inline packages/web-runtime/src/components/Topbar/ThemeSwitcher.vue:27]: Why do we need to cast this? Shouldn't `document.querySelector` return an `HTMLElement` with `style` attribute?

File: packages/design-system/src/components/OcResource/OcResource.vue (hunk 3 above)
Comment (dschmidt) [reply packages/design-system/src/components/OcResource/OcResource.vue:108]: No idea, works fine

File: packages/web-app-files/src/components/Search/List.vue (hunk 8 above)
Comment (dschmidt) [reply packages/web-app-files/src/components/Search/List.vue:172]: Maybe because $route was and is not globally available, let's export it from setup() instead

File: packages/web-app-search/src/views/List.vue (hunk 12 above)
Comment (dschmidt) [reply packages/web-app-search/src/views/List.vue:13]: For some reason vue-tsc thinks `this.$route` is a Function ... I can live with this, we should just port this component to composition api as soon as we touch it again

File: packages/web-runtime/src/components/Topbar/ThemeSwitcher.vue (hunk 14 above)
Comment (dschmidt) [reply packages/web-runtime/src/components/Topbar/ThemeSwitcher.vue:27]: It returns an `Element`, so we need to cast to `HTMLElement`

**Comment by JammingBen** [comment]
LGTM, can't approve though since it's my PR.

File: packages/web-app-files/src/composables/actions/files/useFileActionsDelete.ts [hunk 15]
```diff
@@ -30,7 +30,7 @@ export const useFileActionsDelete = ({ store }: { store?: Store<any> } = {}) =>
       icon: 'delete-bin-5',
       label: () => $gettext('Delete'),
       handler,
-      isEnabled: ({ space, resources }) => {
+      isEnabled: ({ space, resources }: FileActionOptions) => {
```
Comment (dschmidt) [inline packages/web-app-files/src/composables/actions/files/useFileActionsDelete.ts:33]: Unneccessary

File: packages/web-app-files/src/composables/actions/files/useFileActionsDelete.ts [hunk 16]
```diff
@@ -64,7 +64,7 @@ export const useFileActionsDelete = ({ store }: { store?: Store<any> } = {}) =>
...
```
Comment (dschmidt) [inline packages/web-app-files/src/composables/actions/files/useFileActionsDelete.ts:67]: Unneccessary

File: packages/web-pkg/src/components/BatchActions.vue [hunk 17]
```diff
@@ -22,18 +21,19 @@
...
   components: { ActionMenuItem },
   props: {
     items: {
-      type: Array,
+      type: Array as PropType<Resource[]>,
       required: true
     },
     actions: {
-      type: Array,
+      type: Array as PropType<Action[]>,
```
Comment (dschmidt) [inline packages/web-pkg/src/components/BatchActions.vue:36]: Should probably ported to action options as well

Comment (dschmidt) [reply packages/web-pkg/src/components/BatchActions.vue:31]: ... or not as it's generic this way

File: packages/web-app-admin-settings/src/components/General/AppearanceSection.vue [hunk 18]
```diff
@@ -63,6 +63,11 @@ export default defineComponent({
     const menuItems = computed(() =>
       [...instance.$_uploadLogo_items, ...instance.$_resetLogo_items].filter((i) => i.isEnabled())
     )
+
+    const actionOptions = computed(() => ({
+      resources: unref(menuItems)
+    }))
```
Comment (dschmidt) [inline packages/web-app-admin-settings/src/components/General/AppearanceSection.vue:69]: This seems super-odd, but is a 1:1 port from the old code ...

File: packages/web-app-files/src/components/Shares/SharedWithMeSection.vue [hunk 19]
```diff
@@ -48,7 +48,7 @@
             v-if="getShowDeclineButton(resource)"
             size="small"
             class="file-row-share-decline oc-ml-s"
-            @click.stop="triggerAction('decline-share', { resources: [resource] })"
+            @click.stop="triggerAction('decline-share', { space: null, resources: [resource] })"
```
Comment (JammingBen) [inline packages/web-app-files/src/components/Shares/SharedWithMeSection.vue:51]: Why is space`null` here? Shouldn't it be `createShareSpace(resource)` like down below?

File: packages/web-app-files/src/composables/actions/files/useFileActionsSetImage.ts
Comment (JammingBen) [inline packages/web-app-files/src/composables/actions/files/useFileActionsSetImage.ts:1]: This is a space action, why did you move it to the file actions? 🙈

File: packages/web-pkg/src/composables/actions/files/useFileActionsSetReadme.ts
Comment (JammingBen) [inline packages/web-pkg/src/composables/actions/files/useFileActionsSetReadme.ts:1]: This is a space action as well.

File: packages/web-runtime/src/components/Topbar/ApplicationsMenu.vue [hunk 20]
```diff
@@ -53,11 +53,12 @@ import { clientService } from 'web-pkg/src/services'
 import { configurationManager } from 'web-pkg/src/configuration'
 import { mapGetters } from 'vuex'
 import { urlJoin } from 'web-client/src/utils'
+import { PropType } from 'vue-demi'
```
Comment (JammingBen) [inline packages/web-runtime/src/components/Topbar/ApplicationsMenu.vue:56]:
```suggestion
import { PropType } from 'vue'
```

 (probably merge with import above)

File: packages/web-pkg/src/composables/actions/files/useFileActionsSetReadme.ts
Comment (dschmidt) [reply packages/web-pkg/src/composables/actions/files/useFileActionsSetReadme.ts:1]: It's not either - it's an action on files ;-)

File: packages/web-app-files/src/components/Shares/SharedWithMeSection.vue (hunk 19 above)
Comment (dschmidt) [reply packages/web-app-files/src/components/Shares/SharedWithMeSection.vue:51]: Well it's unused in the action handler, so I didn't bother to get a space from somewhere

Comment (JammingBen) [reply packages/web-app-files/src/components/Shares/SharedWithMeSection.vue:51]: As discussed: Let's leave it (for now). The action does not need it and we don't risk any performance impact with hundreds of shares.

Comment (dschmidt) [reply packages/web-app-files/src/components/Shares/SharedWithMeSection.vue:51]: Quickly discussed in chat: We are going to stick with null for now as it's unneeded and it might become a performance problem when having A LOT of spaces. It could be optimized when needed but right now we don't think it makes sense.

**Review by dschmidt — APPROVED** [review]
